### PR TITLE
MotoROS v1.9.2

### PR DESCRIPTION
**Pull request by ted-miller**
- Add validation to the input data for reading/writing I/O.
- Add new result codes to the I/O reply message.
- Add support for reading/writing M-Registers.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
Apologies for leaving this sitting so long @ted-miller. I'll review this Friday.

@steviedale: could you also test this on your controller/robot?

Should make the work needed for #360 much simpler.

**Comment by gavanderhoorn** [comment]
This also seems to address #358.

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/IoServer.c:337]: This seems to now return `0` when `apiRet == OK` in `resultCode`. It used to return `ROS_REPLY_SUCCESS` (which `== 1`).

I'm not necessarily against this change, just wanted to note this somewhere.

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/IoServer.c:338]: Technically, `reply_code` should only encode the success or failure of the reception and processing of the message, not of the execution of the service.

I know this isn't documented very explicitly though. See [this related discussion](https://github.com/ros-industrial/rep/pull/20#discussion_r173632236) with @shaun-edwards.

To summarise:

 - `replyType` should tell clients whether the service request itself was received correctly, and whether the server supports the request.
 - the `resultCode` should then be used to determine whether the action associated with the request was successfully executed.

This is a direct analog of the way ROS services behave.

However, I'm not entirely sure whether we've adhered to this with the other implementations of service servers in MotoROS. Would you know @ted-miller ?

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/IoServer.c:372]: I may have asked this before, but why the `* 10`?

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/IoServer.c:501]: Would it perhaps be better to check this and fail if `1e6` was added by the client?

If I understand the comments in `SimpleMessage.h` correctly, clients should specify the register index directly (ie: 1, 2, 3, etc), instead of adding `1e6` themselves.

For consistency, it may be better to reject `registerNumber > whatever_the_normal_max_is`.

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/IoServer.c]: Is `1` the number of addresses to read? If so: could we avoid the magic nr and use a named constant?

File: motoman_driver/MotoPlus/IoServer.h
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/IoServer.h:61]: These constants seem like they overlap in many places.

Would it be worth it to reuse them if they have the same values? Or would you want to keep the duplicated ones?

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/SimpleMessage.h]: This will require a corresponding update to [REP-I4: Assigned Message Identifiers for the Simple Message Protocol](https://github.com/ros-industrial/rep/blob/bbd88b9b1dc9b61be5178c8b3a93b9bec488702a/rep-I0004.rst), specifically the *Vendor Specific Ranges: Motoman* section.

Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/SimpleMessage.h]: Additionally: there should be no need for separate IDs for requests and replies. Requests and replies share the same ID, so `2012` would be read and `2013` would be write of M registers.

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/SimpleMessage.h:300]: Is this always correct? I'm not completely up-to-speed on how VxWorks maps enums to integer types, but afaik the standard does not necessarily dictate that 4-byte integers must be used for all enums.

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/SimpleMessage.h]:
```suggestion
	UINT32 ioAddress; //Group address. Example: '1001' will read OG#1 (IOs / bits 10010 to 10017)
```
The `-` confused me at first, I thought there was a subtraction there.

Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/SimpleMessage.h]: And the same below.

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]
I've tested this on our YRC1000 and it works. Especially with #371 this results in informative error messages when using the IO interface incorrectly.

It would have been nice to have split this PR into (at least) two commits though: the addition of M-register read/write seems to be fairly stand-alone functionality.

**Comment by gavanderhoorn** [comment]
Thanks for the PR @ted-miller :+1: 

I've left some in-line comments.

File: motoman_driver/MotoPlus/IoServer.c
Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.c:338]: > However, I'm not entirely sure whether we've adhered to this with the other implementations of service servers in MotoROS. Would you know @ted-miller ?

This has not been adhered to throughout MotoROS. Generally speaking, if anything fails, the replyType will not indicate 'success'. Here's an example in the motion server:
https://github.com/ros-industrial/motoman/blob/4b0a030de3917aa13ab1407f2dcf131d4f0ffd27/motoman_driver/MotoPlus/MotionServer.c#L608-L636

File: motoman_driver/MotoPlus/IoServer.c
Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.c:372]: When reading a whole 'group', it is expecting the group-number, which is the first four digits of an I/O address.
For example, for the signal 12345, it is Group 1234, Bit 5.

File: motoman_driver/MotoPlus/IoServer.c
Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.c:501]: > If I understand the comments in SimpleMessage.h correctly, clients should specify the register index directly (ie: 1, 2, 3, etc), instead of adding 1e6 themselves.

Yes. But, I put this catch in there in case someone doesn't read the comment. 
I think it's better to allow for more flexible input format, rather than return an error code and force someone to waste time debugging just because they didn't read a comment in a header file. This extra check eliminates confusion and headache in the future.

File: motoman_driver/MotoPlus/IoServer.c
Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.c]: But magic is fun...

File: motoman_driver/MotoPlus/IoServer.h
Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.h:61]: I don't understand your comment. Can you elaborate?

If you're referring the usage of the controller-model constant, this is something that is defined in the build system for the MotoPlus SDK.

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.c:338]: Hm. That's not nice, but also not something we can change at this point I believe.

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (ted-miller) [reply motoman_driver/MotoPlus/SimpleMessage.h]: I'm fine with removing the 'reply' message ID's. However, this is the method that is also used for motion-control messages.
https://github.com/ros-industrial/motoman/blob/4b0a030de3917aa13ab1407f2dcf131d4f0ffd27/motoman_driver/MotoPlus/SimpleMessage.h#L60-L61

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.c:501]: > I think it's better to allow for more flexible input format, rather than return an error code and force someone to waste time debugging just because they didn't read a comment in a header file. 

that's not the only option we have: we can create an appropriate result_code and return it to clients.

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.c]: Certainly :)

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (ted-miller) [reply motoman_driver/MotoPlus/SimpleMessage.h:300]: I'm not sure if that is dictated by the build system or the OS. But, I can confirm with certainty that an enum is a 4 byte int in this environment.

File: motoman_driver/MotoPlus/IoServer.h
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.h:61]: Example: `IFPANELMIN` and `IFPANELMAX` seem to be the same across all supported controller models.

Instead of defining the values 5 times, we could also do it once, and only provide `#define`s for the values which are different.

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/SimpleMessage.h]: > However, this is the method that is also used for motion-control messages.

I know, but that was also unnecessary.

I'm not sure why/how that got through the reviews at the time.

File: motoman_driver/MotoPlus/IoServer.h
Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.h:61]: Yeah, I originally tried to determine what all the differences were. But, it started to get confusing and difficult to read the code due to all the similarities/differences between controllers.

I think it's easier to read if they are all grouped nicely in a block in numerical order. It will also be easier to update in the future, if this code is around for the next controller generation.

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (ted-miller) [reply motoman_driver/MotoPlus/SimpleMessage.h]: I guess my only concern is breaking existing functionality.

There are users who are currently using the I/O interface, today. I assume that their code will be looking for the "_REPLY" message ID. If we remove these messages, will it break existing installations? (assuming they upgrade their MotoROS at some point)

https://github.com/ros-industrial/motoman/blob/4b0a030de3917aa13ab1407f2dcf131d4f0ffd27/motoman_driver/MotoPlus/SimpleMessage.h#L63-L71

Thoughts?

File: motoman_driver/MotoPlus/IoServer.h
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.h:61]: Ok. Makes sense.

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/SimpleMessage.h]: > I assume that their code will be looking for the "_REPLY" message ID.

could you elaborate how? The numbers you refer to are numeric IDs of message structures. I would expect code to work with the numbers, but not necessarily (de)serialise to different classes/structures.

I'm not suggesting we remove the ones which are already there. It's unfortunate we have those, but removing them may break code outside of this repository.

I'm suggesting to not introduce *more* unnecessary ID allocations.

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/SimpleMessage.h:300]: Hm. I don't personally like to assume things like this (as they tend to change and break things while I'm not looking), but if you're comfortable with this, then we can keep it.

File: motoman_driver/MotoPlus/IoServer.c
Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.c]: This has been updated

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (ted-miller) [reply motoman_driver/MotoPlus/SimpleMessage.h]: This has been updated. Please note that I also changed the value for ROS_MSG_MOTO_WRITE_MREGISTER. So, that will need to be updated on the motomand-driver side too.

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (ted-miller) [reply motoman_driver/MotoPlus/SimpleMessage.h]: This has been updated.

File: motoman_driver/MotoPlus/SimpleMessage.h
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/SimpleMessage.h]: Support for those things need to be added still, so there is nothing to update yet.

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/IoServer.c]: Should this be updated as well to use the constant(s)?

Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.c]: You're right. I just missed that.

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.c:501]: Would returning an error result code be an option @ted-miller? Seems anything out of the range [0, 1000)` would be illegal.

Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.c:501]: I don't see any reason to return an error. If the value is between 0 - 999 or 1000000 - 1000999, then we know what they're trying to do. We might as well honor the request message. There's no benefit to making it more difficult for the user.

Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.c:501]: Reason I suggest returning an error is because it's very valuable to have interfaces which do one thing well, without possibility for ambiguity or allowing unneeded variation.

Allowing both `[0, 999]` and `[1000000, 1000999]` is such unneeded variation. Only by reading documentation would users know which ranges are accepted, and it would only be the documentation here iiuc, as the MotoPlus documentation actually only allows `[1000000, 1000999]`.

Seeing as that's a likely source of information for users / customers of Yaskawa robots, it would seem good to actually only accept `[1000000, 1000999]` and nothing else.

This would make it consistent with available documentation and the the rest of of the IO services we expose here.

Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.c:501]: But I don't feel strongly enough about it to hold up this PR, so if you'd like this merged as-is now, we can do that.

**Comment by gavanderhoorn** [comment]
@steviedale @marip8: would you have a chance to test this on your controller before the end of this week?

It's OK of course if you can't, just let me know and I'll try to test it myself.

**Comment by steviedale** [comment]
@gavanderhoorn Sorry for the delayed response regarding the testing. I won't be in the office until Monday. I can test it first thing on Monday if that isn't too late?

File: motoman_driver/MotoPlus/IoServer.c
Comment (ted-miller) [reply motoman_driver/MotoPlus/IoServer.c:501]: I've had too many occurrences of supporting a user over the course of multiple days, only to find that the root cause was a simple ID or value being misinterpreted. It rightfully upsets the user, and wastes both our time.

Anything that makes an interface more accepting so that it "just works" is a good thing. There is no downside to making something easier for different users. (Especially since this interface is being separated from the motoman-driver so that it can be used independently.)

Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.c:501]: Ok. Let's agree to disagree. Personally I find clear error messages and clearly defined interfaces a good thing, especially if those interfaces adhere to what is documented (in documentation available elsewhere), but I can see how what you describe can be valuable.

**Comment by gavanderhoorn** [comment]
I've tested the YRC1000 binary on our controller today.

Everything seems to be in order.

As such, I'm merging this.

CI is not cooperating due to some changes at Travis. I'll use admin privileges to ignore that and merge the PR.

**Comment by gavanderhoorn** [comment]
Note: the code for reading/writing M-registers has not been tested by me.

That will have to wait until we have client code in `motoman_driver`.

I do not want to block this PR on that becoming available.

**Review by gavanderhoorn — APPROVED** [review]

**Comment by gavanderhoorn** [comment]
Actually, I'll wait for @marip8's approval of #371.

**Comment by gavanderhoorn** [comment]
Thanks @ted-miller for the changes.

**Pull request merged**
